### ISSUE TITLE
Sincere request for scripts to reproduce the results of 3D de novo molecule generation

### ISSUE THREAD

**Issue by kzhpchen** [acharkq/NExT-Mol#6]
The scripts provided in the README do not seem to be the ones to reproduce the results of 3D de novo molecule generation. Could you please provide the complete args for reproducing the results? More specifically, are the following args correct?
```shell
python llm_train.py  \
    --dataset "QM9" \
    --root dataset/QM92014 \
    --llm_tune full  \
    --filename 'qm9_llm' \
    --generate_eval_epoch 5 \
    --llm_model "acharkq/MoLlama" \
    --rand_smiles restricted \
    --temperature 1.0 \
    --num_beams 1 \
    --sample_num 10000 \
    --batch_size 64 \
    --accumulate_grad_batches 2 \
    --max_epochs 100

python train_uncond_gene.py  \
    --dataset "QM9-jodo" \
    --root dataset/QM92014 \
    --filename "qm9_denovo_gen" \
    --llm_model "acharkq/MoLlama" \
    --num_workers 4 \
    --batch_size 128 \
    --max_epochs 1000 \
    --save_every_n_epochs 100 \
    --check_val_every_n_epoch 20 \
    --accumulate_grad_batches 2 \
    --dropout 0.05 \
    --test_conform_epoch 250 \
    --conform_eval_epoch 250 \
    --cache_epoch 2 \
    --generate_eval_epoch 100 \
    --eval_smiles_path NExT-Mol/all_checkpoints/qm9_llm/lightning_logs/version_1/sequences_epoch99.txt \
    --mode train \
    --use_llm \
    --use_flash_attention \
    --ckpt_path ckpt_path_to_model
```

**Comment by lyc0930** [comment]
Hi @SakuraTroyChen , thanks for your interest in our work. To better help you, could you please clarify what specific issue you are encountering?

**Comment by kzhpchen** [comment]
@lyc0930 I am trying to reproduce the results for 3D de novo molecule generation on the QM9 dataset. After running the experiment using the script you provided, the results I obtained do not match those reported in the original paper. I would like to ask if the parameters `--use_llm` and `--ckpt_path` are missing?

**Comment by lyc0930** [comment]
@SakuraTroyChen Could you please share the results you obtained, so we can compare and see where the gap comes from?

**Comment by kzhpchen** [comment]
@lyc0930 Currently, I am temporarily unable to provide the results. I would be very grateful if you could directly provide me with the correct reproduction script and evaluation method. The main mismatch is observed in the FCD of the 3D-metric.

**Comment by acharkq** [comment]
Hi,
Thank you for your interest. We haven’t observed the same issue you mentioned, where only the FCD metric is worse while the others are fine.

From what we see, your script matches ours, and it should reproduce good results even without the LLM. We’ll need more details from your side in order to help debug the issue further.

**Comment by acharkq** [comment]
Also, please be careful not to confuse FCD-3D and FCD-2D—they are different metrics with very different values.

**Comment by kzhpchen** [comment]
@acharkq @lyc0930 My main points of confusion are as follows:
1. Based on the script you provided, during the training of the 3D de novo molecule generation process, is it correct that the LLM is not used, and only the dmt-b is trained and utilized for generation?
2. If question 1 is true, what is the purpose of training the LLM in the first step?
3. If question 1 is false and the LLM is indeed required, should the `--use_self_att_proj` flag also be enabled? And should I load the checkpoint of the LLM trained in Step 1? Additionally, I noticed that even when `use_llm_projector` is set to `False`, the `diffusion_pl.py` script still seems to default to using an operation identical to the `llm_projector`.

**Comment by acharkq** [comment]
**Response to Q1:**

In the first step, molecules are generated using the MolLamma large language model. The second step does not generate new molecules—it only predicts molecular conformers. Therefore, it is fine to omit the LLM in this stage: the DMT model can run independently without the LLM embeddings. On the QM9 dataset, which is a relatively simple dataset, the results should be comparable to those obtained with LLM-based enhancement.

**Response to Q2:**

Training the LLM in the first step produces the generated molecules, stored in the file:
NExT-Mol/all_checkpoints/qm9_llm/lightning_logs/version_1/sequences_epoch99.txt
These molecules can then be used in the second step to predict their conformers.

**Response to Q3:**

For the QM9 dataset, you can simply skip using the LLM in the second step. The diffusion model alone should achieve similar results with those reported in our paper.

**Comment by kzhpchen** [comment]
My issues have been mostly resolved, and I am very grateful for your detailed reply. However, I am still puzzled by the FCD value. My 2D FCD is normal (around 0.07), but my 3D FCD is much lower than 0.8. Could this be related to my RDKit version? What version of RDKit was used for the evaluation? (My version is 2025.3.6).

**Comment by kzhpchen** [comment]
@acharkq @lyc0930 Or, what factors might affect the calculation of 3D FCD? Since I am not an expert in this field, I would like to ask for your guidance. The variables I am currently aware of are:
1. The SMILES txt files generated by the LLM at different training epochs
2. The number of valid SMILES
3. The ChemNet model

At present, I am using the SMILES file generated at epoch 99, and everything else is the same as in the script you provided. However, the 3D FCD value is very small (around 0.2).

**Comment by kzhpchen** [comment]
@acharkq @lyc0930 Due to the failure to reproduce the evaluation metrics, I wanted to test whether the problem lies in the evaluation code. Therefore, I tried to reproduce the results of the "train" row in Table 2(b), but the 3D FCD values are still significantly different (2D FCD = 0.068, 3D FCD = 0.116). Please help me check if there are any issues with the following code:
```shell
file_path=path_to_smiles
eval_mode="3dmol"
llm_model="acharkq/MoLlama"
dataset="QM9-jodo"

python evaluate.py \
    --path $file_path \
    --mode $eval_mode \
    --llm_model $llm_model \
    --dataset $dataset \
    --root path_to_QM92014
```

```python
import argparse
import torch
from rdkit import Chem
from evaluation.eval_functions import get_2D_edm_metric, get_3D_edm_metric
from tqdm import tqdm
from transformers import AutoTokenizer
from data_provider.geom_drugs_jodo_dm import GeomDrugsJODODM
from data_provider.qm9_jodo_dm import QM9DM

def test_valency(molecule):
    has_open_valencies = False
    for atom in molecule.GetAtoms():
        if atom.GetImplicitValence() > 0:
            has_open_valencies = True
            print(True)
            break
    return has_open_valencies



def eval_(args, path, mode):
    tokenizer = AutoTokenizer.from_pretrained(args.llm_model)
    tokenizer.add_bos_token = True
    tokenizer.add_eos_token = True
    if mode == 'smiles':
        if args.dataset == 'QM9-jodo':
            dm = QM9DM(args.root, 2, 64, tokenizer, args)
        elif args.dataset == 'GeomDrugs-JODO':
            dm = GeomDrugsJODODM(args.root, 2, 64, tokenizer, args)
        else:
            raise NotImplementedError(f"dataset {args.dataset} not implemented")
        smiles = load_smiles(path)
        mols = [Chem.MolFromSmiles(smi) for smi in smiles]
        edm2d_dout = get_2D_edm_metric(mols, dm.train_rdmols)

        properties = ['atom_stable', 'mol_stable', 'Complete', 'Unique', 'Novelty']
        print(properties)
        print('\t'.join([str(edm2d_dout[prop]) for prop in properties]))

        print(edm2d_dout)

        moses_out = dm.get_moses_metrics(mols)
        print(moses_out)

        properties = ['SNN', 'Frag', 'Scaf', 'FCD']
        print(properties)
        print('\t'.join([str(moses_out[prop]) for prop in properties]))
    elif mode == '3dmol':
        # mols = torch.load(path)
        from rdkit import RDLogger
        RDLogger.DisableLog('rdApp.*')
        if args.dataset == 'QM9-jodo':
            dm = QM9DM(args.root, 2, 64, tokenizer, args)
            dataset_name = 'QM9'
        elif args.dataset == 'GeomDrugs-JODO':
            dm = GeomDrugsJODODM(args.root, 2, 64, tokenizer, args)
            dataset_name = 'GeomDrugs'
        else:
            raise NotImplementedError(f"dataset {args.dataset} not implemented")

        # if False:
            ## temporary fix on 3d coordinates
            # for mol in tqdm(mols):
            #     conf = mol.GetConformer().GetPositions()
            #     conf = conf / 2.3860 * 2.4777
            #     mol.RemoveAllConformers()
            #     num_atoms = mol.GetNumAtoms()
            #     new_conformer = Chem.Conformer(num_atoms)
            #     for i in range(num_atoms):
            #         new_conformer.SetAtomPosition(i, conf[i].tolist())
            #     mol.AddConformer(new_conformer, assignId=True)


        # train set metrics
        print("######## 2D EDM metric ########")
        sample_mols = random.sample(dm.train_rdmols, 10000)
        edm2d_dout = get_2D_edm_metric(sample_mols, dm.train_rdmols)
        print(edm2d_dout)
        moses_out = dm.get_moses_metrics(sample_mols)
        print(moses_out)
        
        print("######## 3D EDM metric ########")
        eval_results_3d, reconstructed_3d_mols = get_3D_edm_metric(sample_mols, train_mols=dm.train_rdmols, dataset_name=dataset_name)
        print(eval_results_3d)
        eval_results_moses = dm.get_moses_metrics(reconstructed_3d_mols)
        print(eval_results_moses)
        sub_geometry_metric = dm.get_sub_geometry_metric(sample_mols)
        print(sub_geometry_metric)


        # print("######## 2D EDM metric ########")
        # edm2d_dout = get_2D_edm_metric(mols, dm.train_rdmols)
        # print(edm2d_dout)
        # moses_out = dm.get_moses_metrics(mols)
        # print(moses_out)
        
        # print("######## 3D EDM metric ########")
        # eval_results_3d, reconstructed_3d_mols = get_3D_edm_metric(mols, train_mols=dm.train_rdmols, dataset_name=dataset_name)
        # print(eval_results_3d)
        # eval_results_moses =  dm.get_moses_metrics(reconstructed_3d_mols)
        # print(eval_results_moses)
        # sub_geometry_metric = dm.get_sub_geometry_metric(mols)
        # print(sub_geometry_metric)


def load_smiles(path):
    with open(path, 'r') as f:
        lines = f.readlines()
        sampled_sequences = [line.strip().split() for line in lines]
        _, _, smiles = zip(*sampled_sequences)
    return smiles


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--path', type=str, help='Path to the file', default=None)
    parser.add_argument('--mode', type=str, help='Mode for evaluation', default='smiles')
    parser.add_argument('--llm_model', type=str, help='Path to the llm model', default='all_checkpoints/mollama')
    parser.add_argument('--dataset', type=str, help='Path to the llm model', default='GeomDrugs-JODO')
    QM9DM.add_model_specific_args(parser)
    args = parser.parse_args()
    eval_(args, args.path, args.mode)


```

**Comment by kzhpchen** [comment]
@acharkq @lyc0930 
I think I may have found the root of the problem. I randomly sampled 10K training samples from the QM9 dataset preprocessed by JODO and successfully reproduced the performance on the training set using the JODO codebase. However, in the `qm9_dataset_v6.py` used by NExT-MOL, molecules were processed using OpenBabel, which caused a large number of molecules to be discarded due to failures in `Chem.SanitizeMol(mol)`. As a result, reproducing training set performance based on the QM92014 dataset provided by NExT-MOL fails. Not to mention testing the model's performance.

**Is my finding correct? If this is the case, could you provide the correctly preprocessed QM9 dataset, or offer a solution to this issue?**
